core/src/main/java/io/cucumber/core/filter/Filters.java

```diff
@@ -3,49 +3,45 @@
 import gherkin.events.PickleEvent;
 
 import java.net.URI;
-import java.util.ArrayList;
 import java.util.Collection;
 import java.util.List;
 import java.util.Map;
+import java.util.function.Predicate;
 import java.util.regex.Pattern;
 
-public final class Filters {
+public final class Filters implements Predicate<PickleEvent> {
 
-    private final List<PicklePredicate> filters;
+    private Predicate<PickleEvent> filter;
 
     private int count;
 
     public Filters(Options options) {
-        filters = new ArrayList<>();
+        this.filter = t -> true;
         List<String> tagExpressions = options.getTagExpressions();
         if (!tagExpressions.isEmpty()) {
-            this.filters.add(new TagPredicate(tagExpressions));
+            this.filter = this.filter.and(new TagPredicate(tagExpressions));
         }
         List<Pattern> nameFilters = options.getNameFilters();
         if (!nameFilters.isEmpty()) {
-            this.filters.add(new NamePredicate(nameFilters));
+            this.filter = this.filter.and(new NamePredicate(nameFilters));
         }
         Map<URI, ? extends Collection<Integer>> lineFilters = options.getLineFilters();
         if (!lineFilters.isEmpty()) {
-            this.filters.add(new LinePredicate(lineFilters));
+            this.filter = this.filter.and(new LinePredicate(lineFilters));
         }
 
         this.count = options.getLimitCount();
     }
 
-    public boolean matchesFilters(PickleEvent pickleEvent) {
-        for (PicklePredicate filter : filters) {
-            if (!filter.apply(pickleEvent)) {
-                return false;
-            }
-        }
-        return true;
+    @Override
+    public boolean test(PickleEvent pickleEvent) {
+        return this.filter.test(pickleEvent);
     }
 
     public List<PickleEvent> limitPickleEvents(List<PickleEvent> pickleEvents) {
-    	if (count > pickleEvents.size() || count < 1) {
-    		return pickleEvents;
-    	}
-		return pickleEvents.subList(0, count);
-	}
+        if (count > pickleEvents.size() || count < 1) {
+            return pickleEvents;
+        }
+        return pickleEvents.subList(0, count);
+    }
```

Comment:
Ideally where ever `Filters` is used we now use `Predicate<PickleEvent> `.  So might be worth extracting `limitPickleEvents` to some other class.